File: Java/801.java
__________________________________________________________________________________________________
sample 1 ms submission
class Solution {
    public int minSwap(int[] A, int[] B) {
        int swap = 1, keep = 0;
        
        for (int i = 1; i < A.length; i++) {
            if (A[i - 1] >= B[i] || B[i - 1] >= A[i]) {
                swap++;
            }
            else if (A[i - 1] >= A[i] || B[i - 1] >= B[i]) {
                int temp = swap;
                swap = keep + 1;
                keep = temp;
            }
            else {
                int min = Math.min(keep, swap);
                swap = min + 1;
                keep = min;
            }
        }
        
        return Math.min(swap, keep);
    }
}
__________________________________________________________________________________________________
sample 38128 kb submission
class Solution {
    public int minSwap(int[] A, int[] B) {
        memo = new int[A.length][2];
        for (int[] m : memo) {
            Arrays.fill(m, -1);
        }
        return dfs(A, B, 0, 0);
    }
    int memo[][];
    int dfs(int a[], int b[], int pos, int doSwap) {
        if (pos == a.length) {
            return 0;
        }
        if(memo[pos][doSwap] != -1) {
            return memo[pos][doSwap];
        }
        
        int cand1 = Integer.MAX_VALUE;
        int cand2 = Integer.MAX_VALUE;
        if (pos == 0 || a[pos] > a[pos-1] && b[pos] > b[pos-1]) {
            cand1 = dfs(a, b, pos + 1, 0);
        }
        if(pos == 0 || a[pos] > b[pos-1] && b[pos] > a[pos-1]) {
            swap(a, b, pos);
            cand2 = dfs(a,b, pos + 1, 1) + 1;
            swap(a,b,pos);
        }
        
        memo[pos][doSwap] = Math.min(cand1, cand2);
        
        return memo[pos][doSwap];
    }
    public void swap(int[] A,int[] B, int idx){
          int temp=A[idx];
          A[idx]=B[idx];
          B[idx]=temp;
        }    
}
__________________________________________________________________________________________________

File: Python3/801.py
__________________________________________________________________________________________________
sample 88 ms submission
class Solution:
    def minSwap(self, A: List[int], B: List[int]) -> int:
        n = len(A)
        swap, noswap = [n] * n, [n] * n
        swap[0], noswap[0] = 1, 0
        for i in range(1, len(A)):
            if A[i] > A[i-1] and B[i] > B[i-1] and A[i] > B[i-1] and B[i] > A[i-1]:
                noswap[i] = min(noswap[i-1], swap[i-1])
                swap[i] = min(noswap[i-1], swap[i-1]) + 1
            elif A[i] > A[i-1] and B[i] > B[i-1]:
                noswap[i] = noswap[i-1]
                swap[i] = swap[i-1] + 1
            elif A[i] > B[i-1] and B[i] > A[i-1]:
                noswap[i] = swap[i-1]
                swap[i] = noswap[i-1] + 1

        return min(swap[-1], noswap[-1])
__________________________________________________________________________________________________
sample 96 ms submission
class Solution:
    def minSwap(self, A: List[int], B: List[int]) -> int:
        
        minSwap, minNotSwap= 1, 0
        for i in range(1, len(A)):
            swap, notSwap = len(A), len(A)
            
            if A[i] > B[i-1] and B[i] > A[i-1]:
                swap = 1 + minNotSwap
                notSwap = minSwap
            if A[i] > A[i-1] and B[i] > B[i-1]:
                swap = min(swap,1 + minSwap)
                notSwap = min(notSwap, minNotSwap)
            
            minSwap, minNotSwap = swap, notSwap
        
        return min(minSwap, minNotSwap)
__________________________________________________________________________________________________
